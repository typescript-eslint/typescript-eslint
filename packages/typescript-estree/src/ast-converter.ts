import { SourceFile } from 'typescript';
import { convertError, Converter, ASTMaps } from './convert';
import { convertComments } from './convert-comments';
import { convertTokens } from './node-utils';
import { Extra } from './parser-options';
import { TSESTree } from './ts-estree';
import { simpleTraverse } from './simple-traverse';

export function astConverter(
  ast: SourceFile,
  extra: Extra,
  shouldPreserveNodeMaps: boolean,
): { estree: TSESTree.Program; astMaps: ASTMaps | undefined } {
  /**
   * The TypeScript compiler produced fundamental parse errors when parsing the
   * source.
   */
  // internal typescript api...
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const parseDiagnostics = (ast as any).parseDiagnostics;
  if (parseDiagnostics.length) {
    throw convertError(parseDiagnostics[0]);
  }

  /**
   * Recursively convert the TypeScript AST into an ESTree-compatible AST
   */
  const instance = new Converter(ast, {
    errorOnUnknownASTType: extra.errorOnUnknownASTType || false,
    useJSXTextNode: extra.useJSXTextNode || false,
    shouldPreserveNodeMaps,
  });

  const estree = instance.convertProgram();

  /**
   * Optionally remove range and loc if specified
   */
  if (!extra.range || !extra.loc) {
    simpleTraverse(estree, {
      enter: node => {
        if (!extra.range) {
          delete node.range;
        }
        if (!extra.loc) {
          delete node.loc;
        }
      },
    });
  }

  /**
   * Optionally convert and include all tokens in the AST
   */
  if (extra.tokens) {
    estree.tokens = convertTokens(ast);
  }

  /**
   * Optionally convert and include all comments in the AST
   */
  if (extra.comment) {
    estree.comments = convertComments(ast, extra.code);
  }

  const astMaps = shouldPreserveNodeMaps ? instance.getASTMaps() : undefined;

  return { estree, astMaps };
}
