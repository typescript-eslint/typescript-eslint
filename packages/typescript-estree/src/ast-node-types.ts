/**
 * @fileoverview The AST node types produced by the parser.
 * @author Nicholas C. Zakas
 * @author James Henry <https://github.com/JamesHenry>
 * @copyright jQuery Foundation and other contributors, https://jquery.org/
 * MIT License
 */
export const AST_NODE_TYPES: { [key: string]: string } = {
  ArrayExpression: 'ArrayExpression',
  ArrayPattern: 'ArrayPattern',
  ArrowFunctionExpression: 'ArrowFunctionExpression',
  AssignmentExpression: 'AssignmentExpression',
  AssignmentPattern: 'AssignmentPattern',
  AwaitExpression: 'AwaitExpression',
  BigIntLiteral: 'BigIntLiteral',
  BinaryExpression: 'BinaryExpression',
  BlockStatement: 'BlockStatement',
  BreakStatement: 'BreakStatement',
  CallExpression: 'CallExpression',
  CatchClause: 'CatchClause',
  ClassBody: 'ClassBody',
  ClassDeclaration: 'ClassDeclaration',
  ClassExpression: 'ClassExpression',
  ClassImplements: 'ClassImplements',
  ClassProperty: 'ClassProperty',
  ConditionalExpression: 'ConditionalExpression',
  ContinueStatement: 'ContinueStatement',
  DebuggerStatement: 'DebuggerStatement',
  Decorator: 'Decorator',
  DoWhileStatement: 'DoWhileStatement',
  EmptyStatement: 'EmptyStatement',
  ExportAllDeclaration: 'ExportAllDeclaration',
  ExportDefaultDeclaration: 'ExportDefaultDeclaration',
  ExportNamedDeclaration: 'ExportNamedDeclaration',
  ExportSpecifier: 'ExportSpecifier',
  ExpressionStatement: 'ExpressionStatement',
  ForInStatement: 'ForInStatement',
  ForOfStatement: 'ForOfStatement',
  ForStatement: 'ForStatement',
  FunctionDeclaration: 'FunctionDeclaration',
  FunctionExpression: 'FunctionExpression',
  GenericTypeAnnotation: 'GenericTypeAnnotation',
  Identifier: 'Identifier',
  IfStatement: 'IfStatement',
  Import: 'Import',
  ImportDeclaration: 'ImportDeclaration',
  ImportDefaultSpecifier: 'ImportDefaultSpecifier',
  ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
  ImportSpecifier: 'ImportSpecifier',
  JSXAttribute: 'JSXAttribute',
  JSXClosingElement: 'JSXClosingElement',
  JSXClosingFragment: 'JSXClosingFragment',
  JSXElement: 'JSXElement',
  JSXEmptyExpression: 'JSXEmptyExpression',
  JSXExpressionContainer: 'JSXExpressionContainer',
  JSXFragment: 'JSXFragment',
  JSXIdentifier: 'JSXIdentifier',
  JSXMemberExpression: 'JSXMemberExpression',
  JSXNamespacedName: 'JSXNamespacedName',
  JSXOpeningElement: 'JSXOpeningElement',
  JSXOpeningFragment: 'JSXOpeningFragment',
  JSXSpreadAttribute: 'JSXSpreadAttribute',
  JSXSpreadChild: 'JSXSpreadChild',
  JSXText: 'JSXText',
  LabeledStatement: 'LabeledStatement',
  Literal: 'Literal',
  LogicalExpression: 'LogicalExpression',
  MemberExpression: 'MemberExpression',
  MetaProperty: 'MetaProperty',
  MethodDefinition: 'MethodDefinition',
  NewExpression: 'NewExpression',
  ObjectExpression: 'ObjectExpression',
  ObjectPattern: 'ObjectPattern',
  Program: 'Program',
  Property: 'Property',
  RestElement: 'RestElement',
  ReturnStatement: 'ReturnStatement',
  SequenceExpression: 'SequenceExpression',
  SpreadElement: 'SpreadElement',
  Super: 'Super',
  SwitchCase: 'SwitchCase',
  SwitchStatement: 'SwitchStatement',
  TaggedTemplateExpression: 'TaggedTemplateExpression',
  TemplateElement: 'TemplateElement',
  TemplateLiteral: 'TemplateLiteral',
  ThisExpression: 'ThisExpression',
  ThrowStatement: 'ThrowStatement',
  TryStatement: 'TryStatement',
  UnaryExpression: 'UnaryExpression',
  UpdateExpression: 'UpdateExpression',
  VariableDeclaration: 'VariableDeclaration',
  VariableDeclarator: 'VariableDeclarator',
  WhileStatement: 'WhileStatement',
  WithStatement: 'WithStatement',
  YieldExpression: 'YieldExpression',
  /**
   * TS-prefixed nodes
   */
  TSAbstractClassDeclaration: 'TSAbstractClassDeclaration',
  TSAbstractClassProperty: 'TSAbstractClassProperty',
  TSAbstractKeyword: 'TSAbstractKeyword',
  TSAbstractMethodDefinition: 'TSAbstractMethodDefinition',
  TSAnyKeyword: 'TSAnyKeyword',
  TSArrayType: 'TSArrayType',
  TSAsExpression: 'TSAsExpression',
  TSAsyncKeyword: 'TSAsyncKeyword',
  TSBooleanKeyword: 'TSBooleanKeyword',
  TSBigIntKeyword: 'TSBigIntKeyword',
  TSConditionalType: 'TSConditionalType',
  TSConstructorType: 'TSConstructorType',
  TSConstructSignature: 'TSConstructSignature',
  TSDeclareKeyword: 'TSDeclareKeyword',
  TSDeclareFunction: 'TSDeclareFunction',
  TSEnumDeclaration: 'TSEnumDeclaration',
  TSEnumMember: 'TSEnumMember',
  TSExportAssignment: 'TSExportAssignment',
  TSExportKeyword: 'TSExportKeyword',
  TSImportType: 'TSImportType',
  TSInferType: 'TSInferType',
  TSLiteralType: 'TSLiteralType',
  TSIndexedAccessType: 'TSIndexedAccessType',
  TSIndexSignature: 'TSIndexSignature',
  TSInterfaceBody: 'TSInterfaceBody',
  TSInterfaceDeclaration: 'TSInterfaceDeclaration',
  TSInterfaceHeritage: 'TSInterfaceHeritage',
  TSFunctionType: 'TSFunctionType',
  TSMethodSignature: 'TSMethodSignature',
  TSModuleBlock: 'TSModuleBlock',
  TSModuleDeclaration: 'TSModuleDeclaration',
  TSNamespaceFunctionDeclaration: 'TSNamespaceFunctionDeclaration',
  TSNonNullExpression: 'TSNonNullExpression',
  TSNeverKeyword: 'TSNeverKeyword',
  TSNullKeyword: 'TSNullKeyword',
  TSNumberKeyword: 'TSNumberKeyword',
  TSMappedType: 'TSMappedType',
  TSObjectKeyword: 'TSObjectKeyword',
  TSParameterProperty: 'TSParameterProperty',
  TSPrivateKeyword: 'TSPrivateKeyword',
  TSPropertySignature: 'TSPropertySignature',
  TSProtectedKeyword: 'TSProtectedKeyword',
  TSPublicKeyword: 'TSPublicKeyword',
  TSQualifiedName: 'TSQualifiedName',
  TSQuestionToken: 'TSQuestionToken',
  TSReadonlyKeyword: 'TSReadonlyKeyword',
  TSRestType: 'TSRestType',
  TSStaticKeyword: 'TSStaticKeyword',
  TSStringKeyword: 'TSStringKeyword',
  TSSymbolKeyword: 'TSSymbolKeyword',
  TSThisType: 'TSThisType',
  TSTypeAnnotation: 'TSTypeAnnotation',
  TSTypeAliasDeclaration: 'TSTypeAliasDeclaration',
  TSTypeLiteral: 'TSTypeLiteral',
  TSTypeOperator: 'TSTypeOperator',
  TSTypeParameter: 'TSTypeParameter',
  TSTypeParameterDeclaration: 'TSTypeParameterDeclaration',
  TSTypeParameterInstantiation: 'TSTypeParameterInstantiation',
  TSTypePredicate: 'TSTypePredicate',
  TSTypeReference: 'TSTypeReference',
  TSTypeQuery: 'TSTypeQuery',
  TSIntersectionType: 'TSIntersectionType',
  TSTupleType: 'TSTupleType',
  TSOptionalType: 'TSOptionalType',
  TSParenthesizedType: 'TSParenthesizedType',
  TSUnionType: 'TSUnionType',
  TSUndefinedKeyword: 'TSUndefinedKeyword',
  TSUnknownKeyword: 'TSUnknownKeyword',
  TSVoidKeyword: 'TSVoidKeyword'
};
