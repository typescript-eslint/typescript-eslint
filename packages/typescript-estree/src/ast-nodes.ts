import * as es from './ast-tree-nodes';

export type ASTTreeNodes =
  | es.ArrayExpression
  | es.ArrayPattern
  | es.ArrowFunctionExpression
  | es.AssignmentExpression
  | es.AssignmentPattern
  | es.AwaitExpression
  | es.BigIntLiteral
  | es.BinaryExpression
  | es.BlockStatement
  | es.BreakStatement
  | es.CallExpression
  | es.CatchClause
  | es.ClassBody
  | es.ClassDeclaration
  | es.ClassExpression
  | es.ClassProperty
  | es.ConditionalExpression
  | es.ContinueStatement
  | es.DebuggerStatement
  | es.Decorator
  | es.DoWhileStatement
  | es.EmptyStatement
  | es.ExportAllDeclaration
  | es.ExportDefaultDeclaration
  | es.ExportNamedDeclaration
  | es.ExportSpecifier
  | es.ExpressionStatement
  | es.ForInStatement
  | es.ForOfStatement
  | es.ForStatement
  | es.FunctionDeclaration
  | es.FunctionExpression
  | es.Identifier
  | es.IfStatement
  | es.Import
  | es.ImportDeclaration
  | es.ImportDefaultSpecifier
  | es.ImportNamespaceSpecifier
  | es.ImportSpecifier
  | es.JSXAttribute
  | es.JSXClosingElement
  | es.JSXClosingFragment
  | es.JSXElement
  | es.JSXEmptyExpression
  | es.JSXExpressionContainer
  | es.JSXFragment
  | es.JSXIdentifier
  | es.JSXMemberExpression
  // | es.JSXNamespacedName // https://github.com/Microsoft/TypeScript/issues/7411
  | es.JSXOpeningElement
  | es.JSXOpeningFragment
  | es.JSXSpreadAttribute
  | es.JSXSpreadChild
  | es.JSXText
  | es.LabeledStatement
  | es.Literal
  | es.LogicalExpression
  | es.MemberExpression
  | es.MetaProperty
  | es.MethodDefinition
  | es.NewExpression
  | es.ObjectExpression
  | es.ObjectPattern
  | es.Program
  | es.Property
  | es.RestElement
  | es.ReturnStatement
  | es.SequenceExpression
  | es.SpreadElement
  | es.Super
  | es.SwitchCase
  | es.SwitchStatement
  | es.TaggedTemplateExpression
  | es.TemplateElement
  | es.TemplateLiteral
  | es.ThisExpression
  | es.ThrowStatement
  | es.TryStatement
  | es.UnaryExpression
  | es.UpdateExpression
  | es.VariableDeclaration
  | es.VariableDeclarator
  | es.WhileStatement
  | es.WithStatement
  | es.YieldExpression
  | es.TSAbstractClassProperty
  // | es.TSAbstractKeyword
  | es.TSAbstractMethodDefinition
  | es.TSAnyKeyword
  | es.TSArrayType
  | es.TSAsExpression
  | es.TSAsyncKeyword
  | es.TSBooleanKeyword
  | es.TSBigIntKeyword
  | es.TSConditionalType
  | es.TSConstructorType
  | es.TSCallSignatureDeclaration
  | es.TSClassImplements
  | es.TSConstructSignatureDeclaration
  // | es.TSDeclareKeyword
  | es.TSDeclareFunction
  | es.TSEnumDeclaration
  | es.TSEnumMember
  | es.TSExportAssignment
  // | es.TSExportKeyword
  | es.TSExternalModuleReference
  | es.TSImportType
  | es.TSInferType
  | es.TSLiteralType
  | es.TSIndexedAccessType
  | es.TSIndexSignature
  | es.TSInterfaceBody
  | es.TSInterfaceDeclaration
  | es.TSInterfaceHeritage
  | es.TSImportEqualsDeclaration
  | es.TSFunctionType
  | es.TSMethodSignature
  | es.TSModuleBlock
  | es.TSModuleDeclaration
  | es.TSNamespaceExportDeclaration
  | es.TSNonNullExpression
  | es.TSNeverKeyword
  | es.TSNullKeyword
  | es.TSNumberKeyword
  | es.TSMappedType
  | es.TSObjectKeyword
  | es.TSParameterProperty
  | es.TSPrivateKeyword
  | es.TSPropertySignature
  | es.TSProtectedKeyword
  | es.TSPublicKeyword
  | es.TSQualifiedName
  // | es.TSQuestionToken
  // | es.TSReadonlyKeyword
  | es.TSRestType
  | es.TSStaticKeyword
  | es.TSStringKeyword
  | es.TSSymbolKeyword
  | es.TSThisType
  | es.TSTypeAnnotation
  | es.TSTypeAliasDeclaration
  | es.TSTypeAssertion
  | es.TSTypeLiteral
  | es.TSTypeOperator
  | es.TSTypeParameter
  | es.TSTypeParameterDeclaration
  | es.TSTypeParameterInstantiation
  | es.TSTypePredicate
  | es.TSTypeReference
  | es.TSTypeQuery
  | es.TSIntersectionType
  | es.TSTupleType
  | es.TSOptionalType
  | es.TSParenthesizedType
  | es.TSUnionType
  | es.TSUndefinedKeyword
  | es.TSUnknownKeyword
  | es.TSVoidKeyword
  // abstract types
  | es.FunctionDeclarationLike
  | es.ClassPropertyLikeNode
  | es.ClassLikeNode
  | es.TSFunctionDeclarationLike
  | es.TSExpressionWithTypeArgumentsLike;
