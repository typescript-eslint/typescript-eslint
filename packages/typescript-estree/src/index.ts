export * from './parser';
export { ParserServices, TSESTreeOptions } from './parser-options';
export { simpleTraverse } from './simple-traverse';
export * from './ts-estree';
export { clearCaches } from './create-program/createWatchProgram';

// re-export for backwards-compat
export { visitorKeys } from '@typescript-eslint/visitor-keys';

// note - cannot migrate this to an import statement because it will make TSC copy the package.json to the dist folder
export const version: string = require('../package.json').version;
