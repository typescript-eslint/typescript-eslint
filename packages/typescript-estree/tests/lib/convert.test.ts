import type { TSESTree } from '@typescript-eslint/types';
import { AST_NODE_TYPES } from '@typescript-eslint/types';
import * as ts from 'typescript';

import type { TSNode } from '../../src';
import type { ConverterOptions } from '../../src/convert';
import { Converter } from '../../src/convert';

describe('convert', () => {
  function convertCode(code: string): ts.SourceFile {
    return ts.createSourceFile(
      'text.ts',
      code,
      ts.ScriptTarget.ESNext,
      true,
      ts.ScriptKind.TSX,
    );
  }

  it('deeplyCopy should convert node correctly', () => {
    const ast = convertCode('type foo = ?foo<T> | ?(() => void)?');

    function fakeUnknownKind(node: ts.Node): void {
      ts.forEachChild(node, fakeUnknownKind);
      // @ts-expect-error -- intentionally writing to a readonly field
      // eslint-disable-next-line deprecation/deprecation
      node.kind = ts.SyntaxKind.UnparsedPrologue;
    }

    ts.forEachChild(ast, fakeUnknownKind);

    const instance = new Converter(ast);
    expect(instance.convertProgram()).toMatchSnapshot();
  });

  it('deeplyCopy should convert node with decorators correctly', () => {
    const ast = convertCode('@test class foo {}');

    const instance = new Converter(ast);

    expect(
      instance['deeplyCopy'](ast.statements[0] as ts.ClassDeclaration),
    ).toMatchSnapshot();
  });

  it('deeplyCopy should convert node with type parameters correctly', () => {
    const ast = convertCode('class foo<T> {}');

    const instance = new Converter(ast);

    expect(
      instance['deeplyCopy'](ast.statements[0] as ts.ClassDeclaration),
    ).toMatchSnapshot();
  });

  it('deeplyCopy should convert node with type arguments correctly', () => {
    const ast = convertCode('new foo<T>()');

    const instance = new Converter(ast);

    expect(
      instance['deeplyCopy'](
        (ast.statements[0] as ts.ExpressionStatement)
          .expression as ts.NewExpression,
      ),
    ).toMatchSnapshot();
  });

  it('deeplyCopy should convert array of nodes', () => {
    const ast = convertCode('new foo<T>()');

    const instance = new Converter(ast);
    expect(instance['deeplyCopy'](ast)).toMatchSnapshot();
  });

  it('deeplyCopy should fail on unknown node', () => {
    const ast = convertCode('type foo = ?foo<T> | ?(() => void)?');

    const instance = new Converter(ast, {
      errorOnUnknownASTType: true,
    });

    expect(() => instance['deeplyCopy'](ast)).toThrow(
      'Unknown AST_NODE_TYPE: "TSSourceFile"',
    );
  });

  it('nodeMaps should contain basic nodes', () => {
    const ast = convertCode(`
      'test';
      2;
      class foo {};
      type bar = {};
    `);

    const instance = new Converter(ast, {
      shouldPreserveNodeMaps: true,
    });
    instance.convertProgram();
    const maps = instance.getASTMaps();

    function checkMaps(child: ts.SourceFile | ts.Node): void {
      child.forEachChild(node => {
        if (
          node.kind !== ts.SyntaxKind.EndOfFileToken &&
          node.kind !== ts.SyntaxKind.JsxAttributes &&
          node.kind !== ts.SyntaxKind.VariableDeclaration
        ) {
          expect(node).toBe(
            maps.esTreeNodeToTSNodeMap.get(
              maps.tsNodeToESTreeNodeMap.get(node as TSNode),
            ),
          );
        }
        checkMaps(node);
      });
    }

    expect(ast).toBe(
      maps.esTreeNodeToTSNodeMap.get(maps.tsNodeToESTreeNodeMap.get(ast)),
    );
    checkMaps(ast);
  });

  it('nodeMaps should contain jsx nodes', () => {
    const ast = convertCode(`<a.b.c.d.e></a.b.c.d.e>`);

    const instance = new Converter(ast, {
      shouldPreserveNodeMaps: true,
    });
    instance.convertProgram();
    const maps = instance.getASTMaps();

    function checkMaps(child: ts.SourceFile | ts.Node): void {
      child.forEachChild(node => {
        if (
          node.kind !== ts.SyntaxKind.EndOfFileToken &&
          node.kind !== ts.SyntaxKind.JsxAttributes
        ) {
          expect(node).toBe(
            maps.esTreeNodeToTSNodeMap.get(
              maps.tsNodeToESTreeNodeMap.get(node as TSNode),
            ),
          );
        }
        checkMaps(node);
      });
    }

    expect(ast).toBe(
      maps.esTreeNodeToTSNodeMap.get(maps.tsNodeToESTreeNodeMap.get(ast)),
    );
    checkMaps(ast);
  });

  it('nodeMaps should contain export node', () => {
    const ast = convertCode(`export function foo () {}`);

    const instance = new Converter(ast, {
      shouldPreserveNodeMaps: true,
    });
    const program = instance.convertProgram();
    const maps = instance.getASTMaps();

    function checkMaps(child: ts.SourceFile | ts.Node): void {
      child.forEachChild(node => {
        if (node.kind !== ts.SyntaxKind.EndOfFileToken) {
          expect(ast).toBe(
            maps.esTreeNodeToTSNodeMap.get(maps.tsNodeToESTreeNodeMap.get(ast)),
          );
        }
        checkMaps(node);
      });
    }

    expect(ast).toBe(
      maps.esTreeNodeToTSNodeMap.get(maps.tsNodeToESTreeNodeMap.get(ast)),
    );

    expect(maps.esTreeNodeToTSNodeMap.get(program.body[0])).toBeDefined();
    expect(program.body[0]).not.toBe(
      maps.tsNodeToESTreeNodeMap.get(ast.statements[0] as TSNode),
    );
    checkMaps(ast);
  });

  it('should correctly create node with range and loc set', () => {
    const ast = convertCode('');
    const instance = new Converter(ast, {
      shouldPreserveNodeMaps: true,
    });

    const tsNode: ts.KeywordToken<ts.SyntaxKind.AbstractKeyword> = {
      ...ts.factory.createToken(ts.SyntaxKind.AbstractKeyword),
      end: 10,
      pos: 0,
    };
    const convertedNode = instance['createNode'](tsNode, {
      type: AST_NODE_TYPES.TSAbstractKeyword,
      range: [0, 20],
      loc: {
        start: {
          line: 10,
          column: 20,
        },
        end: {
          line: 15,
          column: 25,
        },
      },
    });
    expect(convertedNode).toEqual({
      type: AST_NODE_TYPES.TSAbstractKeyword,
      range: [0, 20],
      loc: {
        start: {
          line: 10,
          column: 20,
        },
        end: {
          line: 15,
          column: 25,
        },
      },
    });
  });

  it('should throw error on jsDoc node', () => {
    const jsDocCode = [
      'const x: function(new: number, string);',
      'const x: function(this: number, string);',
      'var g: function(number, number): number;',
    ];

    for (const code of jsDocCode) {
      const ast = convertCode(code);

      const instance = new Converter(ast);
      expect(() => instance.convertProgram()).toThrow(
        'JSDoc types can only be used inside documentation comments.',
      );
    }
  });

  describe('allowInvalidAST', () => {
    const code = 'const;';

    it(`throws an error for an invalid AST when allowInvalidAST is false`, () => {
      const ast = convertCode(code);

      const instance = new Converter(ast);

      expect(() => instance.convertProgram()).toThrow(
        'A variable declaration list must have at least one variable declarator.',
      );
    });

    it(`does not throw an error for an invalid AST when allowInvalidAST is true`, () => {
      const ast = convertCode(code);

      const instance = new Converter(ast, {
        allowInvalidAST: true,
      });

      expect(() => instance.convertProgram()).not.toThrow();
    });
  });

  describe('suppressDeprecatedPropertyWarnings', () => {
    const getEsCallExpression = (
      converterOptions: ConverterOptions,
    ): TSESTree.CallExpression => {
      const ast = convertCode(`callee<T>();`);
      const tsCallExpression = (ast.statements[0] as ts.ExpressionStatement)
        .expression as ts.CallExpression;
      const instance = new Converter(ast, {
        shouldPreserveNodeMaps: true,
        ...converterOptions,
      });

      instance.convertProgram();

      const maps = instance.getASTMaps();

      return maps.tsNodeToESTreeNodeMap.get(tsCallExpression);
    };

    it('logs on a deprecated property access when suppressDeprecatedPropertyWarnings is false', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation();
      const esCallExpression = getEsCallExpression({
        suppressDeprecatedPropertyWarnings: false,
      });

      // eslint-disable-next-line deprecation/deprecation
      esCallExpression.typeParameters;

      expect(warn).toHaveBeenCalledWith(
        `The 'typeParameters' property is deprecated on CallExpression nodes. Use 'typeArguments' instead. See https://typescript-eslint.io/linting/troubleshooting#the-key-property-is-deprecated-on-type-nodes-use-key-instead-warnings.`,
      );
    });

    it('does not log on a subsequent deprecated property access when suppressDeprecatedPropertyWarnings is false', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation();
      const esCallExpression = getEsCallExpression({
        suppressDeprecatedPropertyWarnings: false,
      });

      /* eslint-disable deprecation/deprecation */
      esCallExpression.typeParameters;
      esCallExpression.typeParameters;
      /* eslint-enable deprecation/deprecation */

      expect(warn).toHaveBeenCalledTimes(1);
    });

    it('does not log on a deprecated property access when suppressDeprecatedPropertyWarnings is true', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation();
      const esCallExpression = getEsCallExpression({
        suppressDeprecatedPropertyWarnings: true,
      });

      // eslint-disable-next-line deprecation/deprecation
      esCallExpression.typeParameters;

      expect(warn).not.toHaveBeenCalled();
    });
  });
});
