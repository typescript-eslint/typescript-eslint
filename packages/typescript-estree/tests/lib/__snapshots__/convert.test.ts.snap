// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert deeplyCopy should convert array of nodes 1`] = `
{
  "amdDependencies": [],
  "bindDiagnostics": [],
  "bindSuggestionDiagnostics": undefined,
  "checkJsDirective": undefined,
  "commentDirectives": undefined,
  "endOfFileToken": {
    "loc": {
      "end": {
        "column": 12,
        "line": 1,
      },
      "start": {
        "column": 12,
        "line": 1,
      },
    },
    "range": [
      12,
      12,
    ],
    "type": "TSEndOfFileToken",
  },
  "externalModuleIndicator": undefined,
  "fileName": "text.ts",
  "hasNoDefaultLib": false,
  "identifierCount": 2,
  "identifiers": Map {
    "foo" => "foo",
    "T" => "T",
  },
  "isDeclarationFile": false,
  "languageVariant": 1,
  "languageVersion": 99,
  "libReferenceDirectives": [],
  "lineMap": [
    null,
  ],
  "loc": {
    "end": {
      "column": 12,
      "line": 1,
    },
    "start": {
      "column": 0,
      "line": 1,
    },
  },
  "nodeCount": 8,
  "parseDiagnostics": [],
  "pragmas": Map {},
  "range": [
    0,
    12,
  ],
  "referencedFiles": [],
  "scriptKind": 4,
  "setExternalModuleIndicator": [Function],
  "statements": [
    {
      "expression": {
        "arguments": [],
        "callee": {
          "loc": {
            "end": {
              "column": 7,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "name": "foo",
          "range": [
            4,
            7,
          ],
          "type": "Identifier",
        },
        "loc": {
          "end": {
            "column": 12,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "range": [
          0,
          12,
        ],
        "type": "NewExpression",
        "typeParameters": {
          "loc": {
            "end": {
              "column": 10,
              "line": 1,
            },
            "start": {
              "column": 7,
              "line": 1,
            },
          },
          "params": [
            {
              "loc": {
                "end": {
                  "column": 9,
                  "line": 1,
                },
                "start": {
                  "column": 8,
                  "line": 1,
                },
              },
              "range": [
                8,
                9,
              ],
              "type": "TSTypeReference",
              "typeName": {
                "loc": {
                  "end": {
                    "column": 9,
                    "line": 1,
                  },
                  "start": {
                    "column": 8,
                    "line": 1,
                  },
                },
                "name": "T",
                "range": [
                  8,
                  9,
                ],
                "type": "Identifier",
              },
              "typeParameters": undefined,
            },
          ],
          "range": [
            7,
            10,
          ],
          "type": "TSTypeParameterInstantiation",
        },
      },
      "loc": {
        "end": {
          "column": 12,
          "line": 1,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "range": [
        0,
        12,
      ],
      "type": "ExpressionStatement",
    },
  ],
  "text": "new foo<T>()",
  "type": "TSSourceFile",
  "typeReferenceDirectives": [],
}
`;

exports[`convert deeplyCopy should convert node correctly 1`] = `
{
  "body": [
    {
      "loc": {
        "end": {
          "column": 35,
          "line": 1,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "name": {
        "escapedText": "foo",
        "loc": {
          "end": {
            "column": 8,
            "line": 1,
          },
          "start": {
            "column": 5,
            "line": 1,
          },
        },
        "originalKeywordKind": undefined,
        "range": [
          5,
          8,
        ],
        "type": "TSUnparsedPrologue",
      },
      "range": [
        0,
        35,
      ],
      "type": "TSUnparsedPrologue",
      "typeAnnotation": null,
      "typeParameters": null,
    },
  ],
  "loc": {
    "end": {
      "column": 35,
      "line": 1,
    },
    "start": {
      "column": 0,
      "line": 1,
    },
  },
  "range": [
    0,
    35,
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`convert deeplyCopy should convert node with decorators correctly 1`] = `
{
  "decorators": [
    {
      "expression": {
        "loc": {
          "end": {
            "column": 5,
            "line": 1,
          },
          "start": {
            "column": 1,
            "line": 1,
          },
        },
        "name": "test",
        "range": [
          1,
          5,
        ],
        "type": "Identifier",
      },
      "loc": {
        "end": {
          "column": 5,
          "line": 1,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "range": [
        0,
        5,
      ],
      "type": "Decorator",
    },
  ],
  "loc": {
    "end": {
      "column": 18,
      "line": 1,
    },
    "start": {
      "column": 0,
      "line": 1,
    },
  },
  "members": [],
  "name": {
    "loc": {
      "end": {
        "column": 15,
        "line": 1,
      },
      "start": {
        "column": 12,
        "line": 1,
      },
    },
    "name": "foo",
    "range": [
      12,
      15,
    ],
    "type": "Identifier",
  },
  "range": [
    0,
    18,
  ],
  "type": "TSClassDeclaration",
  "typeParameters": null,
}
`;

exports[`convert deeplyCopy should convert node with type arguments correctly 1`] = `
{
  "arguments": [],
  "expression": {
    "loc": {
      "end": {
        "column": 7,
        "line": 1,
      },
      "start": {
        "column": 4,
        "line": 1,
      },
    },
    "name": "foo",
    "range": [
      4,
      7,
    ],
    "type": "Identifier",
  },
  "loc": {
    "end": {
      "column": 12,
      "line": 1,
    },
    "start": {
      "column": 0,
      "line": 1,
    },
  },
  "range": [
    0,
    12,
  ],
  "type": "TSNewExpression",
  "typeParameters": {
    "loc": {
      "end": {
        "column": 10,
        "line": 1,
      },
      "start": {
        "column": 7,
        "line": 1,
      },
    },
    "params": [
      {
        "loc": {
          "end": {
            "column": 9,
            "line": 1,
          },
          "start": {
            "column": 8,
            "line": 1,
          },
        },
        "range": [
          8,
          9,
        ],
        "type": "TSTypeReference",
        "typeName": {
          "loc": {
            "end": {
              "column": 9,
              "line": 1,
            },
            "start": {
              "column": 8,
              "line": 1,
            },
          },
          "name": "T",
          "range": [
            8,
            9,
          ],
          "type": "Identifier",
        },
        "typeParameters": undefined,
      },
    ],
    "range": [
      7,
      10,
    ],
    "type": "TSTypeParameterInstantiation",
  },
}
`;

exports[`convert deeplyCopy should convert node with type parameters correctly 1`] = `
{
  "loc": {
    "end": {
      "column": 15,
      "line": 1,
    },
    "start": {
      "column": 0,
      "line": 1,
    },
  },
  "members": [],
  "name": {
    "loc": {
      "end": {
        "column": 9,
        "line": 1,
      },
      "start": {
        "column": 6,
        "line": 1,
      },
    },
    "name": "foo",
    "range": [
      6,
      9,
    ],
    "type": "Identifier",
  },
  "range": [
    0,
    15,
  ],
  "type": "TSClassDeclaration",
  "typeParameters": {
    "loc": {
      "end": {
        "column": 12,
        "line": 1,
      },
      "start": {
        "column": 9,
        "line": 1,
      },
    },
    "params": [
      {
        "constraint": undefined,
        "default": undefined,
        "in": false,
        "loc": {
          "end": {
            "column": 11,
            "line": 1,
          },
          "start": {
            "column": 10,
            "line": 1,
          },
        },
        "name": {
          "loc": {
            "end": {
              "column": 11,
              "line": 1,
            },
            "start": {
              "column": 10,
              "line": 1,
            },
          },
          "name": "T",
          "range": [
            10,
            11,
          ],
          "type": "Identifier",
        },
        "out": false,
        "range": [
          10,
          11,
        ],
        "type": "TSTypeParameter",
      },
    ],
    "range": [
      9,
      12,
    ],
    "type": "TSTypeParameterDeclaration",
  },
}
`;
